app/js/view/upload_view.js

```diff
@@ -0,0 +1,208 @@
+import { IconHelper } from 'js/lib/helpers';
+import ListModel from 'js/model/list_model';
+
+export default class UploadView extends View {
+  constructor() {
+    this.el = document.createElement('div');
+    this.el.id = 'uploads';
+    this.el.classList.add('popup');
+    this.currentApp = null;
+
+    this.uploadBlacklist = [
+      'app://ee50b34d-c8ce-b941-8020-5bc4693c770c/manifest.webapp',//FxStumbler
+      'https://webmaker-app.mofostaging.net/manifest.webapp',
+      'app://a9c7eb67-c2c8-8845-b150-5dd0a850200c/manifest.webapp',//Hackerplace
+      'app://7a6c01c5-32fd-d041-9d0c-a70bb6c7c752/manifest.webapp',
+      'app://customizer.gaiamobile.org/manifest.webapp',
+      'app://default_theme.gaiamobile.org/manifest.webapp'
+    ];
+    // Add current directory apps to upload blacklist.
+    var directoryApps = (new ListModel()).getAppList();
+    for (var manifestURL in directoryApps) {
+      this.uploadBlacklist.push(manifestURL);
+    }
+  }
+
+  render() {
+    super();
+    this.closeButton = this.el.querySelector('.close');
+    this.cancelButton = this.el.querySelector('#upload-cancel');
+    this.submitButton = this.el.querySelector('#upload-submit');
+    this.list = this.el.querySelector('#upload-list');
+    this.displayName = this.el.querySelector('#display-name');
+    this.displayIcon = this.el.querySelector('#display-icon');
+    this.nameInput = this.el.querySelector('#upload-name');
+    this.descriptionInput = this.el.querySelector('#upload-description');
+    this.alertDialog = document.body.querySelector('#alert-dialog');
+
+    this.closeButton.addEventListener('click', this.hide.bind(this));
+    this.cancelButton.addEventListener('click', this.hideForm.bind(this));
+    this.submitButton.addEventListener('click', this.upload.bind(this));
+    this.createList();
+  }
+
+  showAlertDialog(msg) {
+    this.alertDialog.textContent = msg;
+    this.alertDialog.open();
+  }
+
+  isEligible(app) {
+    if (this.uploadBlacklist.indexOf(app.manifestURL) !== -1) {
+      return false;
+    }
+    if (app.manifest.role === 'theme' || app.manifest.role === 'addon') {
+      return true;
+    }
+    // Only non-gaia non-marketplace apps are eligible for hackerplace.
+    return (!app.installOrigin.endsWith('gaiamobile.org') &&
+            app.installOrigin !== 'https://marketplace.firefox.com');
+  }
+
+  createList() {
+    var req;
+    if (navigator.mozApps.mgmt) {
+      req = navigator.mozApps.mgmt.getAll();
+    } else {
+      req = navigator.mozApps.getInstalled();
+    }
+    req.onsuccess = () => {
+      var apps = req.result;
+      apps.forEach(app => {
+        if (this.isEligible(app)) {
+          var item = document.createElement('li');
+          var icon = this.getIconUrl(app.manifest, app.origin);
+          item.classList.add('item');
+          item.innerHTML = this.itemTemplate(app.manifest);
+          IconHelper.setImage(item.querySelector('.icon'), icon);
```

Comment:
This is very fragile. Setting ```innerHTML``` may take one frame. It might be better to wrap this in a ```setTimeout(() => { });``` call.